### PR TITLE
[CALCITE-2601] Add REVERSE function

### PR DESCRIPTION
**Pull request by ambition119**
Fix [ISSUE #2601](https://issues.apache.org/jira/browse/CALCITE-2601)

mysql

```sql
mysql> SELECT REVERSE('hello');
+------------------+
| REVERSE('hello') |
+------------------+
| olleh            |
+------------------+
1 row in set (0.00 sec)
```

sql server

```

DECLARE @str NVARCHAR(100) 

SET @str='ABCD'

SELECT REVERSE(@str)

```

pg

```
testdb=# SELECT REVERSE('abcd');
 reverse
---------
 dcba
(1 row)
```

oracle

```

SQL> select reverse('12345') from dual;
REVER

54321

```

doc:

https://dev.mysql.com/doc/refman/5.7/en/string-functions.html#function_reverse

### CODE REVIEW

**Comment by chunweilei** [comment]
Build failed?

**Comment by ambition119** [comment]
> Build failed?

yes , DocumentationTest failed, I add reference.md content.

**Comment by chunweilei** [comment]
BTW, can you add some tests for UTF-8 like `reverse('你好')`?

**Comment by ambition119** [comment]
> BTW, can you add some tests for UTF-8 like `reverse('你好')`?

done.

**Comment by chunweilei** [comment]
Please address julian's comments:
>It's not a standard function. Can you create a new function category "j" (for JDBC).
And move it to SqlLibraryOperators.
Also change the JIRA case title to "Add REVERSE function"; whether it is in SqlFunctions is an implementation detail.
In the description, you don't need to say that it returns null if the input is null.

**Comment by ambition119** [comment]
> Please address julian's comments:
> 
> > It's not a standard function. Can you create a new function category "j" (for JDBC).
> > And move it to SqlLibraryOperators.
> > Also change the JIRA case title to "Add REVERSE function"; whether it is in SqlFunctions is an implementation detail.
> > In the description, you don't need to say that it returns null if the input is null.

done, thanks.

**Comment by ambition119** [comment]
@chunweilei @julianhyde 
I fix the issue comment code, and commit build success.  But PR 1 failing check, so I can comment 'recheck' then recheck all the code?
企业微信截图_18af6e36-07c4-4d90-ad76-6a79fed792e7

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java [hunk 1]
```diff
@@ -1501,6 +1501,16 @@ public SqlOperandCountRange getOperandCountRange() {
           OperandTypes.CHARACTER,
           SqlFunctionCategory.STRING);
 
+  public static final SqlFunction REVERSE =
+      new SqlFunction(
+          "REVERSE",
+          SqlKind.REVERSE,
+          ReturnTypes.ARG0_NULLABLE_VARYING,
+          null,
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: Cause it is not standard, maybe we  should not add the `REVERSE` implementation in `SqlStdOperatorTable`, move it directly to `SqlLibraryOperators` is more acceptable.

File: site/_docs/reference.md [hunk 2]
```diff
@@ -2108,6 +2109,7 @@ semantics.
 | o | GREATEST(expr [, expr ]*)                      | Returns the greatest of the expressions
 | o | LEAST(expr [, expr ]* )                        | Returns the least of the expressions
 | o p | TRANSLATE(expr, fromString, toString)          | Returns *expr* with all occurrences of each character in *fromString* replaced by its corresponding character in *toString*. Characters in *expr* that are not in *fromString* are not replaced
+| m | REVERSE(string)                                  | Returns the reverse order of string values
```
Comment (danny0405) [inline site/_docs/reference.md]: Move this item together with other mysql functions.

Comment (ambition119) [reply site/_docs/reference.md]: done.

File: core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java (hunk 1 above)
Comment (ambition119) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: > Cause it is not standard, maybe we should not add the `REVERSE` implementation in `SqlStdOperatorTable`, move it directly to `SqlLibraryOperators` is more acceptable.

thank you review this pr, I fix these code.

Comment (ambition119) [reply core/src/main/java/org/apache/calcite/sql/fun/SqlStdOperatorTable.java]: before PR code like:
```java
@LibraryOperator(libraries = {MYSQL})
  public static final SqlFunction JSON_PRETTY = SqlStdOperatorTable.JSON_PRETTY;
```

**Comment by ambition119** [comment]
@julianhyde 
I don't see the details of the check error, I can only see the code that can't pull the PR, so I can't help now.

By the way, in [Apache pulsar](https://github.com/apache/pulsar/pull/4281), we can re-checks the PR by commenting on keywords like `rerun java8 tests`. I hope that calcite can also handle this, thank you!

**Comment by chunweilei** [comment]
Please:
* fix build fail
* squash commits to one
* change the commit message to '[CALCITE-2601] Add REVERSE function ( your jira id)'

**Comment by ambition119** [comment]
> Please:
> 
> * fix build fail

I PR `mvn clean install` build success, now I can observe that I have submitted the failed display, and 
**AppVeyor**  did not pull the code to **check** and **build**, thank you.

> * squash commits to one

done.

> * change the commit message to '[CALCITE-2601] Add REVERSE function ( your jira id)'

done.

**Comment by ambition119** [comment]
Is conflicts lead `AppVeyor was unable to build `?  If I resolved the conflicts, then it’s not good to go to squash commits to one.  thanks

**Comment by chunweilei** [comment]
The error says `AppVeyor was unable to build non-mergable pull request`. So you should resolve conflicts first. BTW, you commit message is `LCITE-2601] Add REVERSE function`. Please correct it.

**Comment by ambition119** [comment]
> The error says `AppVeyor was unable to build non-mergable pull request`. So you should resolve conflicts first. BTW, you commit message is `LCITE-2601] Add REVERSE function`. Please correct it.

thanks

File: site/_docs/reference.md [hunk 3]
```diff
@@ -2097,6 +2098,7 @@ semantics.
 | m | JSON_LENGTH(jsonValue [, path ])               | Returns a integer indicating the length of *jsonValue*
 | m | JSON_KEYS(jsonValue [, path ])                 | Returns a string indicating the keys of a JSON *jsonValue*
 | m | JSON_REMOVE(jsonValue, path[, path])           | Removes data from *jsonValue* using a series of *path* expressions and returns the result
+| m | REVERSE(string)                                | Returns the reverse order of string values
 | o | DECODE(value, value1, result1 [, valueN, resultN ]* [, default ]) | Compares *value* to each *valueN* value one by one; if *value* is equal to a *valueN*, returns the corresponding *resultN*, else returns *default*, or NULL if *default* is not specified
```
Comment (chunweilei) [inline site/_docs/reference.md]: Please change the description to `Returns the reverse order of *string*` just like other format.

**Comment by chunweilei** [comment]
Also please resolve conflicts and squash commits.

**Comment by ambition119** [comment]
> Also please resolve conflicts and squash commits.

Thank you for your reminders and suggestions, but because I am merge apache/master into origin/reverse_fun, so now squash commits is more troublesome, these commit code enter this PR.
```shell
pick 62cf3d1dc [CALCITE-2601] Add REVERSE function
pick d0a06f786 [CALCITE-3061] Query with WITH clause fails when alias is the same as the table with rolled up column
pick c25b29c0c [CALCITE-2803] ProjectTransposeJoinRule messes INDF expressions
pick 2a2140f98 Site: Add Danny Chan as committer
pick b039a36a3 [CALCITE-3034] CSV test case description does not match it's code logic (FaxianZhao)
pick 3f24710db [CALCITE-2936] Simplify EXISTS or NOT EXISTS sub-query that has "GROUP BY ()"
pick e98c779d1 [CALCITE-3017] Improve null handling of JsonValueExpressionOperator
pick a5b382abb [CALCITE-3028] Support FULL OUTER JOIN with AggregateJoinTransposeRule (Vineet Garg)
pick d6896202c [CALCITE-3066] RelToSqlConverter may incorrectly throw an AssertionError for some decimal literals
pick 046bb81a9 [CALCITE-3052] Error while applying rule MaterializedViewAggregateRule(Project-Aggregate): ArrayIndexOutOfBoundsException
pick 051058be8 [CALCITE-2282] Remove sql operator table from parser
pick 7df06784a [CALCITE-2933] Add timestamp extract for casts from timestamp type to other types
pick 999c41d5f [CALCITE-2975] Implement JSON_REMOVE function (xuqianjin)
pick c6b6800c2 [CALCITE-3062] Do not populate provenanceMap if not debug
pick ca92ce27c [CALCITE-3074] Move MySQL's JSON operators to SqlLibraryOperators
pick 4d0477305 [CALCITE-2965] Implement string functions: REPEAT, SPACE, SOUNDEX, DIFFERENCE
pick cc73fe525 merge conflicts
pick 020c60bad fix doc
```
By the way, this is not my first PR, The code that has been merged is https://github.com/apache/calcite/blob/master/server/src/main/java/org/apache/calcite/sql/ddl/SqlCreateFunction.java

**Comment by chunweilei** [comment]
@ambition119 you should use `--rebase` when you try to merge master to your branch. I will try to merge this PR.

**Comment by ambition119** [comment]
> @ambition119 you should use `--rebase` when you try to merge master to your branch. I will try to merge this PR.

thanks, I use 
```
$ git rebase -i HEAD~4
```
then merge master,  but the commit(like pick xx fix doc) after the merge is difficult to `--rebase`.

**Comment by chunweilei** [comment]
You can checkout a new branch from master and cherry-pick some commits from this branch. Then delete this branch and rename the new branch to this branch name. Finally, you force push the new branch.

**Comment by ambition119** [comment]
> You can checkout a new branch from master and cherry-pick some commits from this branch. Then delete this branch and rename the new branch to this branch name. Finally, you force push the new branch.

Thank you for your advice, I will try and understand this development process. My usual development operation is
```shell
1. git clone https://github.com/<Github_user>/calcite.git
2. $ git remote add apache https://github.com/apache/calcite
3. $ git fetch apache
    $ git rebase apache/master
    $ git push   
4. $ git checkout -b reverse_fun
```
then dev and commit,  and squash commits is 
```shell
$ git rebase -i HEAD~5
$ git push --force
```
Can you add your WeChat?   my wechat is 358998716, thanks.

**Comment by chunweilei** [comment]
You almost do the right thing. I think you may wrong only when you resolve conflicts. You should:
```
git checkout master
git fetch apache
git merge apache/master
git push origin master
git checkout dev_branch
git pull --rebase origin master
```

**Comment by chunweilei** [comment]
@ambition119 could you please review https://github.com/chunweilei/calcite/commit/b9c3e33b7174941e8cf0c25af29430114f50fce4? If no other comments, I would like to merge it.

**Comment by ambition119** [comment]
> @ambition119 could you please review [chunweilei@b9c3e33](https://github.com/chunweilei/calcite/commit/b9c3e33b7174941e8cf0c25af29430114f50fce4)? If no other comments, I would like to merge it.

Wait a moment, I submit a test code for you to see. Thank you！

**Comment by ambition119** [comment]
> @ambition119 could you please review [chunweilei@b9c3e33](https://github.com/chunweilei/calcite/commit/b9c3e33b7174941e8cf0c25af29430114f50fce4)? If no other comments, I would like to merge it.

@chunweilei Thank you for your help and processing,  the [chunweilei@b9c3e33](https://github.com/chunweilei/calcite/commit/b9c3e33b7174941e8cf0c25af29430114f50fce4) is no problem, just added the add reverse jdbc test in testJdbcFn(), thank you!

**Pull request closed without merging**